File: Lumo/Theme/Theme.swift
import SwiftUI
import UIKit

/// Oyunun tüm renk kimliği tek yerden yönetilir.
/// Her tema hem SwiftUI (Color) hem SpriteKit (UIColor) tarafında kullanılır.
struct ThemeColor: Equatable {
    let r: Double
    let g: Double
    let b: Double

    var color: Color { Color(red: r, green: g, blue: b) }
    var uiColor: UIColor { UIColor(red: r, green: g, blue: b, alpha: 1) }

    func opacity(_ a: Double) -> Color { Color(red: r, green: g, blue: b).opacity(a) }
}

struct Theme: Identifiable, Equatable {
    let id: String
    /// İngilizce ad — aynı zamanda yerelleştirme anahtarıdır (Localizable.xcstrings)
    let name: String
    let isPremium: Bool

    var localizedName: String { String(localized: String.LocalizationValue(name)) }

    let bgTop: ThemeColor
    let bgBottom: ThemeColor
    let ring: ThemeColor
    let gate: ThemeColor
    let orb: ThemeColor
    let hazard: ThemeColor
    let lumen: ThemeColor
    let accent: ThemeColor

    /// Renk körlüğü modunda üretilmiş bir tema mı? Oyun alanı bu bayrağa
    /// bakarak renge ek olarak ŞEKİL ipucu da çiziyor (tehlike yaylarındaki
    /// tırtıklar), çünkü tek başına renk hiçbir palette yeterli değil.
    var isColorBlindSafe = false

    /// "Bu tur yakmaz" durumundaki tehlike yayının rengi. Temanın `accent`
    /// rengi kullanılamıyor: bazı temalarda (mercan) vurgu, tehlikenin
    /// kendisine çok yakın bir somon tonu — yay silahlı mı değil mi
    /// anlaşılmıyor. Tüm temaların halkası soğuk (mor/mavi/yeşil/beyaz)
    /// olduğu için magenta hem halkayla hem kırmızıyla karışmıyor.
    var hazardSafe: ThemeColor {
        isColorBlindSafe
        ? ThemeColor(r: 0.800, g: 0.475, b: 0.655)   // Okabe–Ito kırmızımsı mor
        : ThemeColor(r: 0.839, g: 0.361, b: 0.941)   // magenta
    }

    // Varsayılan tema. Arka plan bilerek DÜŞÜK doygunlukta: eski mor
    // (0.129, 0.043, 0.267) uzun oturumlarda göz yoruyordu. Kimlik aynı
    // kaldı, sadece morun kanı çekildi ve biraz maviye kaydırıldı.
    static let nebula = Theme(
        id: "nebula", name: "Nebula", isPremium: false,
        bgTop: ThemeColor(r: 0.035, g: 0.033, b: 0.068),
        bgBottom: ThemeColor(r: 0.070, g: 0.057, b: 0.112),
        ring: ThemeColor(r: 0.689, g: 0.661, b: 0.939),
        gate: ThemeColor(r: 0.290, g: 0.949, b: 0.788),
        orb: ThemeColor(r: 1.0, g: 1.0, b: 1.0),
        hazard: ThemeColor(r: 1.0, g: 0.302, b: 0.416),
        lumen: ThemeColor(r: 1.0, g: 0.827, b: 0.353),
        accent: ThemeColor(r: 0.639, g: 0.545, b: 1.0)
    )

    static let gece = Theme(
        id: "gece", name: "Night", isPremium: false,
        bgTop: ThemeColor(r: 0.017, g: 0.031, b: 0.055),
        bgBottom: ThemeColor(r: 0.031, g: 0.073, b: 0.117),
        ring: ThemeColor(r: 0.546, g: 0.791, b: 0.967),
        gate: ThemeColor(r: 0.478, g: 1.0, b: 0.643),
        orb: ThemeColor(r: 0.918, g: 0.976, b: 1.0),
        hazard: ThemeColor(r: 1.0, g: 0.275, b: 0.263),
        lumen: ThemeColor(r: 1.0, g: 0.851, b: 0.400),
        accent: ThemeColor(r: 0.353, g: 0.784, b: 1.0)
    )

    // Pembe/gül tonlu tema — ama halka SOĞUK (periwinkle) olduğu için
    // kırmızı tehlike net ayrışır, göz yormaz.
    static let safak = Theme(
        id: "safak", name: "Dawn", isPremium: true,
        bgTop: ThemeColor(r: 0.081, g: 0.045, b: 0.111),
        bgBottom: ThemeColor(r: 0.219, g: 0.095, b: 0.164),
        ring: ThemeColor(r: 0.776, g: 0.819, b: 1.000),
        gate: ThemeColor(r: 1.0, g: 0.871, b: 0.549),
        orb: ThemeColor(r: 1.0, g: 0.965, b: 0.933),
        hazard: ThemeColor(r: 1.0, g: 0.302, b: 0.290),
        lumen: ThemeColor(r: 1.0, g: 0.816, b: 0.302),
        accent: ThemeColor(r: 1.0, g: 0.620, b: 0.549)
    )

    // Orman: halka yeşil olduğundan kapı AÇIK CAMGÖBEĞİ — sarı yıldızlarla
    // ve yeşil halkayla karışmaz; kırmızı tehlike yeşilin tam karşıtıdır.
    static let orman = Theme(
        id: "orman", name: "Forest", isPremium: true,
        bgTop: ThemeColor(r: 0.019, g: 0.061, b: 0.052),
        bgBottom: ThemeColor(r: 0.044, g: 0.125, b: 0.095),
        ring: ThemeColor(r: 0.628, g: 0.921, b: 0.724),
        gate: ThemeColor(r: 0.451, g: 0.902, b: 1.0),
        orb: ThemeColor(r: 0.949, g: 1.0, b: 0.949),
        hazard: ThemeColor(r: 1.0, g: 0.290, b: 0.235),
        lumen: ThemeColor(r: 1.0, g: 0.851, b: 0.400),
        accent: ThemeColor(r: 0.478, g: 0.918, b: 0.643)
    )

    // Mercan/okyanus teması — halka AQUA (soğuk), kapı amber, tehlike canlı
    // mercan-kırmızısı: üçü de birbirinden net ayrışır.
    static let mercan = Theme(
        id: "mercan", name: "Coral", isPremium: true,
        bgTop: ThemeColor(r: 0.027, g: 0.052, b: 0.087),
        bgBottom: ThemeColor(r: 0.037, g: 0.139, b: 0.168),
        ring: ThemeColor(r: 0.546, g: 0.901, b: 0.903),
        gate: ThemeColor(r: 1.0, g: 0.804, b: 0.451),
        orb: ThemeColor(r: 1.0, g: 0.976, b: 0.949),
        hazard: ThemeColor(r: 1.0, g: 0.271, b: 0.302),
        lumen: ThemeColor(r: 1.0, g: 0.851, b: 0.451),
        accent: ThemeColor(r: 1.0, g: 0.549, b: 0.502)
    )

    // Aurora: yıldızlar tüm temalarda olduğu gibi AMBER (öğretici "sarıları
    // topla" der — yeşil yıldız kafa karıştırıyordu); tehlike net kırmızı.
    static let aurora = Theme(
        id: "aurora", name: "Aurora", isPremium: true,
        bgTop: ThemeColor(r: 0.026, g: 0.018, b: 0.066),
        bgBottom: ThemeColor(r: 0.027, g: 0.099, b: 0.120),
        ring: ThemeColor(r: 0.579, g: 0.956, b: 0.842),
        gate: ThemeColor(r: 0.788, g: 0.510, b: 1.0),
        orb: ThemeColor(r: 0.949, g: 1.0, b: 0.988),
        hazard: ThemeColor(r: 1.0, g: 0.263, b: 0.310),
        lumen: ThemeColor(r: 1.0, g: 0.851, b: 0.400),
        accent: ThemeColor(r: 0.427, g: 0.949, b: 0.800)
    )

    // Yüksek kontrastlı premium tema — neredeyse siyah zeminde elektrik renkleri;
    // her öğe (özellikle kırmızı tehlike) keskin ayrışır.
    static let neon = Theme(
        id: "neon", name: "Neon", isPremium: true,
        bgTop: ThemeColor(r: 0.013, g: 0.013, b: 0.033),
        bgBottom: ThemeColor(r: 0.037, g: 0.022, b: 0.065),
        ring: ThemeColor(r: 0.480, g: 0.936, b: 1.000),
        gate: ThemeColor(r: 0.451, g: 1.0, b: 0.549),
        orb: ThemeColor(r: 1.0, g: 1.0, b: 1.0),
        hazard: ThemeColor(r: 1.0, g: 0.153, b: 0.353),
        lumen: ThemeColor(r: 1.0, g: 0.902, b: 0.251),
        accent: ThemeColor(r: 0.851, g: 0.353, b: 1.0)
    )

    // Ultra yüksek kontrast / erişilebilirlik teması — beyaz halka, saf kırmızı
    // tehlike, yeşil kapı (trafik ışığı netliği); kömür siyahı zemin.
    static let karbon = Theme(
        id: "karbon", name: "Carbon", isPremium: true,
        bgTop: ThemeColor(r: 0.024, g: 0.024, b: 0.030),
        bgBottom: ThemeColor(r: 0.064, g: 0.064, b: 0.073),
        ring: ThemeColor(r: 0.967, g: 0.967, b: 1.000),
        gate: ThemeColor(r: 0.302, g: 1.0, b: 0.502),
        orb: ThemeColor(r: 0.549, g: 0.847, b: 1.0),
        hazard: ThemeColor(r: 1.0, g: 0.200, b: 0.200),
        lumen: ThemeColor(r: 1.0, g: 0.800, b: 0.200),
        accent: ThemeColor(r: 0.600, g: 0.651, b: 0.749)
    )

    // Lüks tema: gece mavisi kadife zemin, platin halka, zümrüt kapı,
    // altın vurgu — "kraliyet" havası, ama oyun okunurluğu tam.
    static let kraliyet = Theme(
        id: "kraliyet", name: "Royal", isPremium: true,
        bgTop: ThemeColor(r: 0.025, g: 0.033, b: 0.076),
        bgBottom: ThemeColor(r: 0.057, g: 0.070, b: 0.145),
        ring: ThemeColor(r: 0.860, g: 0.890, b: 0.967),
        gate: ThemeColor(r: 0.318, g: 0.902, b: 0.647),
        orb: ThemeColor(r: 1.0, g: 0.988, b: 0.949),
        hazard: ThemeColor(r: 1.0, g: 0.251, b: 0.290),
        lumen: ThemeColor(r: 1.0, g: 0.804, b: 0.361),
        accent: ThemeColor(r: 0.949, g: 0.780, b: 0.416)
    )

    // Sakura: mürekkep koyuluğunda erik zemin, lavanta halka, nane kapı,
    // kiraz çiçeği vurgusu — pembe yalnızca süslemede, tehlike net kırmızı.
    static let sakura = Theme(
        id: "sakura", name: "Sakura", isPremium: true,
        bgTop: ThemeColor(r: 0.063, g: 0.036, b: 0.070),
        bgBottom: ThemeColor(r: 0.131, g: 0.069, b: 0.118),
        ring: ThemeColor(r: 0.819, g: 0.784, b: 1.000),
        gate: ThemeColor(r: 0.471, g: 0.949, b: 0.737),
        orb: ThemeColor(r: 1.0, g: 0.965, b: 0.976),
        hazard: ThemeColor(r: 1.0, g: 0.239, b: 0.251),
        lumen: ThemeColor(r: 1.0, g: 0.831, b: 0.400),
        accent: ThemeColor(r: 1.0, g: 0.678, b: 0.796)
    )

    static let all: [Theme] = [.nebula, .gece, .safak, .orman, .mercan, .aurora,
                               .neon, .karbon, .kraliyet, .sakura]

    static func theme(id: String) -> Theme {
        all.first { $0.id == id } ?? .nebula
    }

    // MARK: Renk körlüğü modu

    /// Oyunun anlamı renge bağlı: yeşil kapı "git", kırmızı yay "ölürsün",
    /// sarı yıldız "topla". En yaygın renk körlüğü tam bu kırmızı–yeşil
    /// ayrımını siliyor. Bu mod açıkken zemin ve arka plan temanın kendisi
    /// olarak kalıyor, ama OYNANIŞI belirleyen dört renk Okabe–Ito
    /// paletinden sabit değerlerle değiştiriliyor: bu palet protanopi,
    /// döteranopi ve tritanopide de birbirinden ayrışır.
    ///
    /// Halka da nötr griye çekiliyor; yoksa kapının mavisi bazı temaların
    /// mavi halkasıyla karışıyor.
    var colorBlindSafe: Theme {
        Theme(
            id: id, name: name, isPremium: isPremium,
            bgTop: bgTop, bgBottom: bgBottom,
            ring: ThemeColor(r: 0.788, g: 0.808, b: 0.863),      // nötr gri-mavi
            gate: ThemeColor(r: 0.196, g: 0.588, b: 0.894),      // doygun mavi
            orb: ThemeColor(r: 1.0, g: 1.0, b: 1.0),
            hazard: ThemeColor(r: 0.902, g: 0.400, b: 0.055),    // vermilyon
            lumen: ThemeColor(r: 0.941, g: 0.894, b: 0.259),     // sarı
            accent: ThemeColor(r: 0.0, g: 0.694, b: 0.506),      // mavimsi yeşil
            isColorBlindSafe: true
        )
    }
}

File: android/app/src/main/java/com/caganhatapci/orbeon/theme/Theme.kt
package com.caganhatapci.orbeon.theme

import androidx.compose.ui.graphics.Color
import com.caganhatapci.orbeon.R

/** Oyunun tüm renk kimliği tek yerden yönetilir. iOS sürümüyle aynı paletler. */
data class Theme(
    val id: String,
    /** Yerelleştirme kaynak kimliği */
    val nameRes: Int,
    val isPremium: Boolean,
    val bgTop: Color,
    val bgBottom: Color,
    val ring: Color,
    val gate: Color,
    val orb: Color,
    val hazard: Color,
    val lumen: Color,
    val accent: Color,
    /**
     * Renk körlüğü modunda üretilmiş bir tema mı? Oyun alanı bu bayrağa
     * bakarak renge ek olarak ŞEKİL ipucu da çiziyor (tehlike yaylarındaki
     * çentikler), çünkü tek başına renk hiçbir palette yeterli değil.
     */
    val isColorBlindSafe: Boolean = false
) {
    /**
     * "Bu tur yakmaz" durumundaki tehlike yayının rengi. Temanın `accent`
     * rengi kullanılamıyor: bazı temalarda (mercan) vurgu, tehlikenin
     * kendisine çok yakın bir somon tonu — yay silahlı mı değil mi
     * anlaşılmıyor. Tüm temaların halkası soğuk (mor/mavi/yeşil/beyaz)
     * olduğu için magenta hem halkayla hem kırmızıyla karışmıyor.
     */
    val hazardSafe: Color
        get() = if (isColorBlindSafe) Color(0.800f, 0.475f, 0.655f, 1f)
                else Color(0.839f, 0.361f, 0.941f, 1f)

    /**
     * Oyunun anlamı renge bağlı: yeşil kapı "git", kırmızı yay "ölürsün",
     * sarı yıldız "topla". En yaygın renk körlüğü tam bu kırmızı–yeşil
     * ayrımını siliyor. Bu mod açıkken zemin temanın kendisi olarak kalıyor,
     * ama OYNANIŞI belirleyen renkler Okabe–Ito paletinden sabit değerlerle
     * değiştiriliyor: bu palet protanopi, döteranopi ve tritanopide de
     * birbirinden ayrışır. Halka da nötr griye çekiliyor; yoksa kapının
     * mavisi bazı temaların mavi halkasıyla karışıyor.
     */
    val colorBlindSafe: Theme
        get() = copy(
            ring = Color(0.788f, 0.808f, 0.863f, 1f),      // nötr gri-mavi
            gate = Color(0.196f, 0.588f, 0.894f, 1f),      // doygun mavi
            orb = Color(1f, 1f, 1f, 1f),
            hazard = Color(0.902f, 0.400f, 0.055f, 1f),    // vermilyon
            lumen = Color(0.941f, 0.894f, 0.259f, 1f),     // sarı
            accent = Color(0f, 0.694f, 0.506f, 1f),        // mavimsi yeşil
            isColorBlindSafe = true
        )

    companion object {
        private fun c(r: Double, g: Double, b: Double) =
            Color(r.toFloat(), g.toFloat(), b.toFloat(), 1f)

        // Varsayılan tema. Arka plan bilerek DÜŞÜK doygunlukta: eski mor
        // (0.129, 0.043, 0.267) uzun oturumlarda göz yoruyordu.
        val nebula = Theme("nebula", R.string.theme_nebula, false,
            c(0.035,0.033,0.068), c(0.070,0.057,0.112), c(0.689,0.661,0.939),
            c(0.290,0.949,0.788), c(1.0,1.0,1.0), c(1.0,0.302,0.416),
            c(1.0,0.827,0.353), c(0.639,0.545,1.0))

        val gece = Theme("gece", R.string.theme_night, false,
            c(0.017,0.031,0.055), c(0.031,0.073,0.117), c(0.546,0.791,0.967),
            c(0.478,1.0,0.643), c(0.918,0.976,1.0), c(1.0,0.275,0.263),
            c(1.0,0.851,0.400), c(0.353,0.784,1.0))

        val safak = Theme("safak", R.string.theme_dawn, true,
            c(0.081,0.045,0.111), c(0.219,0.095,0.164), c(0.776,0.819,1.000),
            c(1.0,0.871,0.549), c(1.0,0.965,0.933), c(1.0,0.302,0.290),
            c(1.0,0.816,0.302), c(1.0,0.620,0.549))

        val orman = Theme("orman", R.string.theme_forest, true,
            c(0.019,0.061,0.052), c(0.044,0.125,0.095), c(0.628,0.921,0.724),
            c(0.451,0.902,1.0), c(0.949,1.0,0.949), c(1.0,0.290,0.235),
            c(1.0,0.851,0.400), c(0.478,0.918,0.643))

        val mercan = Theme("mercan", R.string.theme_coral, true,
            c(0.027,0.052,0.087), c(0.037,0.139,0.168), c(0.546,0.901,0.903),
            c(1.0,0.804,0.451), c(1.0,0.976,0.949), c(1.0,0.271,0.302),
            c(1.0,0.851,0.451), c(1.0,0.549,0.502))

        val aurora = Theme("aurora", R.string.theme_aurora, true,
            c(0.026,0.018,0.066), c(0.027,0.099,0.120), c(0.579,0.956,0.842),
            c(0.788,0.510,1.0), c(0.949,1.0,0.988), c(1.0,0.263,0.310),
            c(1.0,0.851,0.400), c(0.427,0.949,0.800))

        val neon = Theme("neon", R.string.theme_neon, true,
            c(0.013,0.013,0.033), c(0.037,0.022,0.065), c(0.480,0.936,1.000),
            c(0.451,1.0,0.549), c(1.0,1.0,1.0), c(1.0,0.153,0.353),
            c(1.0,0.902,0.251), c(0.851,0.353,1.0))

        val karbon = Theme("karbon", R.string.theme_carbon, true,
            c(0.024,0.024,0.030), c(0.064,0.064,0.073), c(0.967,0.967,1.000),
            c(0.302,1.0,0.502), c(0.549,0.847,1.0), c(1.0,0.200,0.200),
            c(1.0,0.800,0.200), c(0.600,0.651,0.749))

        val kraliyet = Theme("kraliyet", R.string.theme_royal, true,
            c(0.025,0.033,0.076), c(0.057,0.070,0.145), c(0.860,0.890,0.967),
            c(0.318,0.902,0.647), c(1.0,0.988,0.949), c(1.0,0.251,0.290),
            c(1.0,0.804,0.361), c(0.949,0.780,0.416))

        val sakura = Theme("sakura", R.string.theme_sakura, true,
            c(0.063,0.036,0.070), c(0.131,0.069,0.118), c(0.819,0.784,1.000),
            c(0.471,0.949,0.737), c(1.0,0.965,0.976), c(1.0,0.239,0.251),
            c(1.0,0.831,0.400), c(1.0,0.678,0.796))

        val all = listOf(nebula, gece, safak, orman, mercan, aurora,
                         neon, karbon, kraliyet, sakura)

        fun byId(id: String): Theme = all.firstOrNull { it.id == id } ?: nebula
    }
}
